quic-connection-dos/src/main.rs

```diff
@@ -0,0 +1,369 @@
+
+    endpoint.set_default_client_config(config);
+
+    endpoint
+}
+
+pub fn get_client_config(keypair: &Keypair) -> ClientConfig {
+    let ipaddr = IpAddr::V4(Ipv4Addr::LOCALHOST);
+    let (cert, key) = new_self_signed_tls_certificate(keypair, ipaddr)
+        .expect("Failed to generate client certificate");
+
+    let mut crypto = rustls::ClientConfig::builder()
+        .with_safe_defaults()
+        .with_custom_certificate_verifier(SkipServerVerification::new())
+        .with_single_cert(vec![cert], key)
+        .expect("Failed to use client certificate");
+
+    crypto.enable_early_data = true;
+    crypto.alpn_protocols = vec![ALPN_TPU_PROTOCOL_ID.to_vec()];
+
+    let mut config = ClientConfig::new(Arc::new(crypto));
+
+    let mut transport_config = TransportConfig::default();
+    let timeout = IdleTimeout::try_from(QUIC_MAX_TIMEOUT).unwrap();
+    transport_config.max_idle_timeout(Some(timeout));
+    transport_config.keep_alive_interval(Some(QUIC_KEEP_ALIVE));
+    config.transport_config(Arc::new(transport_config));
+
+    config
+}
+
+// Opens a slow stream and tries to send PACKET_DATA_SIZE bytes of junk
+// in as many chunks as possible. We don't allow the number of chunks
+// to be configurable as client-side writes don't correspond to
+// quic-level packets/writes (but by doing multiple writes we are generally able
+// to get multiple writes on the quic level despite the spec not guaranteeing this)
+pub async fn check_multiple_writes(conn: &Connection) {
+    // Send a full size packet with single byte writes.
+    let num_bytes = PACKET_DATA_SIZE;
+    let mut s1 = conn.open_uni().await.unwrap();
+    for _ in 0..num_bytes {
+        s1.write_all(&[0u8]).await.unwrap();
+    }
+    s1.finish().await.unwrap();
+}
+
+async fn run_connection_dos(
+    server_address: SocketAddr,
+    num_connections: u64,
+    num_streams_per_conn: u64,
+) {
+    let (cert, priv_key) =
+        new_self_signed_tls_certificate(&Keypair::new(), IpAddr::V4(Ipv4Addr::UNSPECIFIED))
+            .expect("Failed to create cert");
+
+    let cert = Arc::new(QuicClientCertificate {
+        certificate: cert,
+        key: priv_key,
+    });
+
+    let endpoint = create_endpoint(cert);
+
+    let mut connections = vec![];
+    for _ in 0..num_connections {
+        connections.push(
+            endpoint
+                .connect(server_address, "connect")
+                .expect("Failed in connecting")
+                .await
+                .expect("Failed in waiting"),
+        );
+    }
+
+    let futures: Vec<_> = connections
+        .iter()
+        .map(|conn| (0..num_streams_per_conn).map(|_| check_multiple_writes(conn)))
+        .flatten()
+        .collect();
+
+    join_all(futures).await;
+}
+
+fn main() {
+    solana_logger::setup();
+
+    let matches = App::new(crate_name!())
+        .about(crate_description!())
+        .version(solana_version::version!())
+        .arg(
+            Arg::with_name("target_address")
```

Comment:
Do we use name `target_address` somewhere else? I feel like entrypoint is used, maybe semantics is different

Reply:
Not sure - what name would you prefer?